drivers/net/mlx5/mlx5_vdpa.c

```diff
@@ -147,14 +160,92 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
 	 * For now only set protocol features support
 	 */
 	priv->caps.max_num_virtqs = MLX5_VDPA_SW_MAX_VIRTQS_SUPPORTED;
-	priv->caps.virtio_net_features = (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+	/*
+	 * TODO (shahafs): Take from QUERY HCA CAP Device Emulation Capabilities.
+	 * For now only set protocol features support
+	 */
+	priv->caps.notify_stride = MLX5_VDPA_SW_NOTIFY_STRIDE;
+	priv->caps.virtio_net_features = MLX5_VDPA_FEATURES;
+	priv->caps.virtio_protocol_features = MLX5_VDPA_PROTOCOL_FEATURES;
 	DRV_LOG(DEBUG, "Virtio Caps:");
 	DRV_LOG(DEBUG, "	dump_mkey=0x%x ", priv->caps.dump_mkey);
 	DRV_LOG(DEBUG, "	max_num_virtqs=0x%x ", priv->caps.max_num_virtqs);
 	DRV_LOG(DEBUG, "	features_bits=0x%" PRIx64, priv->caps.virtio_net_features);
 	return 0;
 }
 
+#define MLX5_IB_MMAP_CMD_SHIFT 8
+#define MLX5_IB_MMAP_CMD_MASK ((1 << MLX5_IB_MMAP_CMD_SHIFT) - 1)
+#define MLX5_IB_CMD_SIZE 8
+#define MLX5_IB_MMAP_VIRTIO_NOTIFY 9
+static inline int
+mlx5_vdpa_get_notify_offset(uint16_t offset)
+{
+	uint16_t ext_offset = MLX5_IB_MMAP_CMD_SHIFT + MLX5_IB_CMD_SIZE;
+	return (((offset >> MLX5_IB_MMAP_CMD_SHIFT) << ext_offset) |
+		(MLX5_IB_MMAP_VIRTIO_NOTIFY << MLX5_IB_MMAP_CMD_SHIFT) |
+		(offset & MLX5_IB_MMAP_CMD_MASK));
```

Comment:
This function is not readable enough, very hard to read this piece of code and understand what it does.
Please consider making it more simple and layered, like in libmlx5, for instance.

offset = 0;
set_command(MLX5_IB_MMAP_VIRTIO_NOTIFY, &offset);
set_extended_index(offset, &offset)
.
.
.

Reply:
yes, i didn't think that it justify to have such comprehensive APIs only for setting a single command. 

i can change it, no problem.